joomla-projects/gsoc21_core-enhancements
# Display Modules Position Badge Background Color based on the Active Template Positions

Pull Request for Issue #2 

### Summary of Changes

Added a new model `getValidPositions` that performs the following:
1. Fetch the active template from the database (1 Query)
2. Get the `templateDetails.xml` file corresponding to the template received as a result of Step 1
3. Store these positions in an array and return it

The array returned from the above model method will be used in the `_getList` function as:
1. Stored the result of the `getValidPositions`  method in an array
2. After fetching all modules from the database for rendering, perform a check whether each module row's position exists in the array of step 1
        If `yes`, set `$row->activePosition` to `true`
        If `no`, set `$row->activePosition` to `false`

Finally, in the `tmpl` file, we use this `activePosition` flag to conditionally select between `bg-primary` and `bg-secondary`


Extra:
Added a few missing template positions to the XML file:
For example, Atum's cpanel-system:
![image](https://user-images.githubusercontent.com/53610833/122663364-79589180-d1b7-11eb-83d4-286e380ea5b1.png)


<hr>

### Testing Instructions
Visit the database and change the value of one of the module's positions to a random string.
 
**Note**: Positions for multi-lingual website could be an edge case but I am unable to set up a multi-lingual site on my local workstation so I'm unable to confirm this

<hr>

### Actual result BEFORE applying this Pull Request

All badges have the same `bg-primary` class


<hr>

### Expected result AFTER applying this Pull Request

Badges for In-active (not of the active template) Positions have a `bg-secondary` class

![image](https://user-images.githubusercontent.com/53610833/122663117-d2272a80-d1b5-11eb-8a94-57e00c94507b.png)



<hr>

### Documentation Changes Required
I don't know
